beacon_node/beacon_chain/src/beacon_chain.rs

```diff
@@ -441,6 +441,47 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
             .map(|result| result.map_err(|e| e.into())))
     }
 
+    /// Iterate through the current chain to find the slot intersecting with the given beacon state.
+    /// The maximum depth this will search is `SLOTS_PER_HISTORICAL_ROOT`, and if that depth is reached
+    /// and no intersection is found, the finalized slot will be returned.
+    pub fn find_reorg_slot(
+        &self,
+        new_state: &BeaconState<T::EthSpec>,
+        new_block_root: &Hash256,
+    ) -> Result<Slot, Error> {
+        self.with_head(|snapshot| {
+            let slot = snapshot.beacon_block.slot();
+
+            // Check the head block
+            if &snapshot.beacon_block_root == new_block_root {
+                return Ok(slot);
+            }
+
+            let next_slot = slot - Slot::new(1);
+            for (i, _) in snapshot.beacon_state.block_roots.iter().enumerate() {
+                let ancestor_slot = next_slot - Slot::new(i as u64);
+                let ancestor_block_root = snapshot.beacon_state.get_block_root(ancestor_slot)?;
```

Comment:
I spent quite a while trying to find edge-cases with the genesis state and forwards/backwards reorgs but I couldn't make anything stick.

One thing I noticed is that this line contains an implicit assumption that the two blocks share the same genesis block (or else an error is returned). This is a totally reasonable assumption, but the implicit assumption without a comment looked like an attack vector. 

I even wrote another implementation during this journey, to help me understand the problem better: https://github.com/paulhauner/lighthouse/commit/f82d2b139d1daa409830661f6cb361b6ff7516b2

I'm really not convinced that my implementation is better than yours though. It's potentially more explicit, but it's more LoC.

I think I might sit on this over the weekend and come back to it with a fresh mind. I'd also be interested in the fresh eyes of @michaelsproul, if he has the time/desire.

Reply:
I'm a big fan of the pair of iterators in your implementation. I think it makes it way simpler conceptually. I cherry picked that commit and pushed it, let me know if you'd like me to revert it 🙁